File: go/backtracking/784_letter_case_permutation.go
package backtracking

import (
	"strings"
	"unicode"
)

// letterCasePermutation 字母大小写全排列
func letterCasePermutation(s string) []string {
	// 回溯

	if len(s) == 0 {
		return nil
	}

	// 所有结果的合集
	var res = make([]string, 0)
	// 当前查询的路径合集
	path := make([]string, 0)

	// idx 当前遍历到s 的索引
	var backtrack func(idx int)
	backtrack = func(idx int) {

		if idx == len(s) {
			p := strings.Join(path, "")
			res = append(res, p)
			return

		}

		path = append(path, string(s[idx]))
		backtrack(idx + 1)
		path = path[:len(path)-1]

		if s[idx] >= 'a' && s[idx] <= 'z' {
			path = append(path, string(unicode.ToUpper(rune(s[idx]))))
			backtrack(idx + 1)
			path = path[:len(path)-1]
		} else if s[idx] >= 'A' && s[idx] <= 'Z' {
			path = append(path, string(unicode.ToLower(rune(s[idx]))))
			backtrack(idx + 1)
			path = path[:len(path)-1]
		}

	}

	backtrack(0)
	return res
}

File: rust/src/backtracking/_784_letter_case_permutation.rs
/// 字母大小写全排列
pub fn letter_case_permutation(s: String) -> Vec<String> {
    if s.is_empty() {
        return vec![];
    }

    let mut res = vec![];

    let mut path = "".to_string();
    let s: Vec<_> = s.chars().collect();

    backtrack(&s, 0, &mut res, &mut path);
    res
}

fn backtrack(s: &[char], idx: usize, res: &mut Vec<String>, path: &mut String) {
    if idx == s.len() {
        res.push(path.clone());
        return;
    }

    path.push(s[idx]);
    backtrack(s, idx + 1, res, path);
    path.pop();

    if s[idx].is_ascii_lowercase() {
        path.push(s[idx].to_ascii_uppercase());
        backtrack(s, idx + 1, res, path);
        path.pop();
    } else if s[idx].is_ascii_uppercase() {
        path.push(s[idx].to_ascii_lowercase());
        backtrack(s, idx + 1, res, path);
        path.pop();
    }
}
